### ISSUE TITLE
udocker support for github packages?

### ISSUE THREAD

**Issue by katakombi** [indigo-dc/udocker#225]
Hi there,
 recently github started support for hosting docker images for repos.
I can use them in docker:
```
docker pull docker.pkg.github.com/dolmades/dolmades-docker/runtime:1.2
``` 
or in the Dockerfile
`FROM docker.pkg.github.com/dolmades/dolmades-docker/runtime:1.2`

I wonder if it is possible to use those in udocker:
```
udocker pull docker.pkg.github.com/dolmades/dolmades-docker/runtime:1.2
Error: no files downloaded
```
I also played with the `--registry` option but this does not work either:
```
udocker pull --registry=https://docker.pkg.github.com dolmades/dolmades-docker/runtime:1.2
Error: no files downloaded
```
I see there is also an option called `--index` but I don't understand what it does.

If I alter the URL of the registry I get also
`Error: image not found`
so I take this as hint that the URL is correct.

Does someone know what prevents udocker from working with github?

**Comment by jorge-lip** [comment]
Fixes have been added to the devel branch, please check.

**Comment by katakombi** [comment]
From current devel:
```
udocker pull --registry=https://docker.pkg.github.com dolmades/dolmades-docker/runtime:1.2
Error: layers section missing in manifest
Error: no files downloaded
```
Am I using wrong syntax?

**Comment by jorge-lip** [comment]
You need to login with token. Then
udocker pull docker. pkg. github. com/dolmades/dolmades-docker/runtime:1.2

or specify both index and registry as arguments.
Cheers

**Comment by katakombi** [comment]
Sorry to bother again but I don't seem to get the login right. Whenever I use --password, --username, --registry it ignores it. What would be the right syntax to call udocker login?

**Comment by jorge-lip** [comment]
Hi this works for me,

`udocker.py  login  --registry=https://docker.pkg.github.com  --login=USERNAME  --password=TOKEN`

`udocker.py  pull  docker.pkg.github.com/indigo-dc/udocker/test`

Make sure you are using the latest version from devel.

**Comment by katakombi** [comment]
I'm afraid it doesn't work:
```
./udocker login --registry=https://docker.pkg.github.com --login=katakombi --password=TOKEN
username: katakombi
Error: syntax error at: --login=katakombi
```
It keeps asking me for my username. This is devel(head) from your curl cmd.

**Comment by jorge-lip** [comment]
Sorry is not --login is --username

**Comment by katakombi** [comment]
```
./udocker login  --username=katakombi --password=VERYLONGTOKENWORKSWITHDOCKER
Info: installing 1.1.3-3
Info: installation of udockertools successful
stefan@khaleesi:/tmp$ ./udocker pull docker.pkg.github.com/dolmades/dolmades-docker/runtime:1.2
Error: manifest not found or not authorized
Error: no files downloaded
```

**Comment by jorge-lip** [comment]
I don't see packages published under dolmades/dolmades-docker/runtime 
Do you have 1.2 published ?

**Comment by katakombi** [comment]
can you access this link? https://github.com/dolmades/dolmades-docker/packages/22887
I\ve just copypasted the commands for docker
```
docker pull docker.pkg.github.com/dolmades/dolmades-docker/runtime:1.2
```
and those work

**Comment by jorge-lip** [comment]
Ok I see it and it is working for me:

./udocker.py login --registry=docker.pkg.github.com --username=jorge-lip --password=XXXXXXX
./udocker.py  pull docker.pkg.github.com/dolmades/dolmades-docker/runtime:1.2
Downloading layer: sha256:7e1eb4178be530d773f519ab3008be22fa02e6a13e9dba2d6bfc189c607e2ad2
Downloading layer: sha256:b809a35dfd3612833139cad73db0a5926dd17e035ad2c75d4ff306cbf7ce40f5
Downloading layer: sha256:5f164ab6670990fdb3d56ac75bc8ba4a3c3aa222541fb77d5bdb5ec8a720c922
Downloading layer: sha256:2d4971c608a492592f0faff0fbf83209e1a7c36221d587c900b329d3d325d2d1
Downloading layer: sha256:6001e1789921cf851f6fb2e5fe05be70f482fe9c2286f66892fe5a3bc404569c
Downloading layer: sha256:8e829fe70a46e3ac4334823560e98b257234c23629f19f05460e21a453091e6d
Downloading layer: sha256:251f5509d51d9e4119d4ffb70d4820f8e2d7dc72ad15df3ebd7cd755539e40fd
Downloading layer: sha256:35c102085707f703de2d9eaad8752d6fe1b8f02b5d2149f1d8357c9cc7fb7d0a

Can you please delete your udocker keystore and try again ?
rm  $HOME/.udocker/keystore

**Comment by katakombi** [comment]
Still the same. Can I PM you my access token? I can revoke it once you got it working...

```
./udocker pull docker.pkg.github.com/dolmades/dolmades-docker/runtime:1.2
Error: manifest not found or not authorized
Error: no files downloaded
```

**Comment by katakombi** [comment]
Ooookay, I've created a personal access token with RO permissions and it works now! Seems like I copy&pasted what was in ~/.docker/config.json but this token is not the format which is expected by udocker.

There is another caveat I've run into: once I use https:// and omit it the other time it doesn't use the stored credentials. Actually I could find two entries, maybe udocker could print a warning about that, or if it doesn't actually succeed in logging in.

Sorry for the hassle, and thanks for the support! Udocker is a great tool to me. Closing as of now...
